content/assignments/techjobs-console/_index.md

```diff
@@ -0,0 +1,477 @@
+1. The method parses and filters that data.
+1. The method presents the data in a useful manner.
+
+### The `GetUserSelection()` Method
+
+The `GetUserSelection()` method takes in a string to display above the
+menu, to provide context for what they are being asked. It also takes in
+a `Dictionary` with string keys and string values. How is this used? What
+will this `Dictionary` contain when the method runs?
+
+To figure this out, right-click on the method name and select *Find (All)
+References*. This will open a pane and display each location in the program
+where `GetUserSelection()` is called. The first such usage is the first
+line of the main `while loop`:
+
+```csharp
+   string actionChoice = GetUserSelection("View Jobs", actionChoices);
+```
+
+What is this `Dictionary` named `actionChoices`? If we look a few lines
+above, we see:
+
+```csharp {linenos=true,linenostart=13}
+   // Top-level menu options
+   Dictionary<string, string> actionChoices = new Dictionary<string, string>();
+   actionChoices.Add("search", "Search");
+   actionChoices.Add("list", "List");
+```
+
+If you recall how the program worked when you ran it, the first menu
+that you chose had two options, *Search* and *List*, which seem to
+correspond to the entries in `actionChoices`. This is, in fact, the
+case. This is the data that is used to generate the first menu we see
+when running the program.
+
+The second usage of `GetUserSelection()` is a few lines below:
+
+```csharp {linenos=table,linenostart = 39}
+   string columnChoice = GetUserSelection("List", columnChoices);
+```
+
+This references `columnChoices`, which is declared at the top of
+`RunProgram()` and has a similar structure to `actionChoices` (they're the
+same data type and are used in calls to the same method, so this
+shouldn't be surprising). Most of the entries in `columnChoices`
+correspond to columns in the jobs data set, but there's one additional
+entry with key/value pair `"all"`/ `"All"`. These entries will help
+us present to the user the options for searching our data, which will
+correspond to searching within a given column, or searching all columns
+at once.
+
+The keys in `actionChoices` and `columnChoices` represent the
+*internal* strings we'll use to refer to these options (e.g. when representing
+the user's menu choice, or querying data). The values in the `Dictionary` represent the
+*external* way that these are represented to the user.
+
+Within `GetUserSelection()` itself, most of the code is within a
+`do-while loop`. A [do-while
+loop](https://www.tutorialsteacher.com/csharp/csharp-do-while-loop)
+is similar to a `while` loop, but the conditional check is at the
+*end* of the loop's code block. This has the net consequence that the
+loop's code block *always runs at least once*. At the end of the block's
+execution, we check a condition to determine if we should run the block
+again. This nicely mimics the behavior of simple menu-driven
+applications.
```

Comment:
Have we defined menu-driven applications?

Reply:
They learn about them earlier.

Reply:
Thanks for the clarification.